src/app-layer-htp.c

```diff
@@ -691,99 +603,81 @@ static uint8_t HTPHandleErrorGetId(const char *msg)
  */
 static void HTPHandleError(HtpState *s, const uint8_t dir)
 {
-    if (s == NULL || s->conn == NULL ||
-        s->conn->messages == NULL) {
-        return;
-    }
-
-    size_t size = htp_list_size(s->conn->messages);
-    size_t msg;
-    if(size >= HTP_MAX_MESSAGES) {
-        if (s->htp_messages_offset < HTP_MAX_MESSAGES) {
-            //only once per HtpState
-            HTPSetEvent(s, NULL, dir, HTTP_DECODER_EVENT_TOO_MANY_WARNINGS);
-            s->htp_messages_offset = HTP_MAX_MESSAGES;
-            //too noisy in fuzzing
-            //DEBUG_VALIDATE_BUG_ON("Too many libhtp messages");
-        }
+    if (s == NULL || s->conn == NULL || s->htp_messages_count >= HTP_MAX_MESSAGES) {
         // ignore further messages
         return;
     }
 
-    for (msg = s->htp_messages_offset; msg < size; msg++) {
-        htp_log_t *log = htp_list_get(s->conn->messages, msg);
-        if (log == NULL)
+    htp_log_t *log = htp_conn_next_log(s->conn);
+    while (log != NULL) {
+        char *msg = htp_log_message(log);
+        if (msg == NULL) {
+            htp_log_free(log);
+            log = htp_conn_next_log(s->conn);
             continue;
+        }
 
-        HtpTxUserData *htud = NULL;
-        htp_tx_t *tx = log->tx; // will be NULL in <=0.5.9
-        if (tx != NULL)
-            htud = (HtpTxUserData *) htp_tx_get_user_data(tx);
-
-        SCLogDebug("message %s", log->msg);
+        SCLogDebug("message %s", msg);
 
-        uint8_t id = HTPHandleErrorGetId(log->msg);
-        if (id == 0) {
-            id = HTPHandleWarningGetId(log->msg);
-            if (id == 0)
-                id = HTTP_DECODER_EVENT_UNKNOWN_ERROR;
+        htp_log_code_t id = htp_log_code(log);
+        if (id != HTP_LOG_CODE_UNKNOWN && id != HTP_LOG_CODE_ERROR) {
```

Comment:
Why this check ?

Reply:
Because the existing code doesn't set events for UNKNOWN errors:

https://github.com/OISF/suricata/blob/55c4834e4e9b14a441b735f84d8d35b4eb151702/src/app-layer-htp.c#L732

Because `HTTP_DECODER_EVENT_UNKNOWN_ERROR` maps to `0`. 

This may be a mistranslation of intent in the original code though. If the intent was to avoid setting events after the `strcmp()` thing failed to find an appropriate `HTTP_DECODER_EVENT` to set for the event (due to mismatched libhtp, or whatnot) then we don't have that problem anymore, since the event and the message should always be available now. If the intent was to suppress unknown / other / generic errors for other reasons (useless, noisy, etc.), then this does the same thing.

Reply:
Ok for me